File: examples/guess-w/test_referee.cpp
#include <iostream>
#include <cstdlib>
#include <string>
#include <vector>

#include "socketstream.h"
#include "player_connection.h"

void display_results(std::string result_type,
		     net::socketstream& manager,
		     std::vector<PlayerConnection>& players)
{
    manager << result_type << ":end" << std::endl;

    for (auto player : players) {
	manager << result_type << "result:";
	manager << player.get_results() << std::endl;
    }
}

void run_game(net::socketstream& manager,
	      std::vector<PlayerConnection>& players)
{
    manager << "round:start|{}" << std::endl;
    for (auto& player : players) {
	player.clear_won();
    }

    std::string buffer;
    std::string winner;
    bool done = false;
    for (int i = 0; !done; ++i) {
	for (auto& player : players) {
	    player.sendline("move");

	    buffer = player.getline();

	    manager << "move: turn " << i << " is " << buffer;
	    manager << "|{player: \"" << player.get_name() << "\", ";
	    manager << "value: \"" << buffer << "\", ";
	    manager << "win: \"" << (buffer == "w") << "\"}" << std::endl;

	    if (buffer == "w") {
		winner = player.get_name();
		player.set_won();
		player.set_score(player.get_score() + 1.0);
		done = true;
		break;
	    }
	}
    }

    for (auto player : players) {
	player.sendline(winner + " wins!");
    }

    display_results("round", manager, players);
}

int main(int argc, char **argv) {
    net::socketstream manager("localhost", std::atoi(argv[1]));
    net::server referee;

    manager << "port:" << referee.get_port() << std::endl;
    manager << "match:start" << std::endl;

    int num_players = std::atoi(argv[2]);
    std::vector<PlayerConnection> players;
    for (int i = 0; i < num_players; ++i) {
	players.push_back(PlayerConnection(referee.serve()));
    }

    for (int i = 0; i < std::atoi(argv[3]); ++i) {
	run_game(manager, players);
    }

    display_results("match", manager, players);

    return 0;
}

File: examples/guess-w/test_referee.py
#! /usr/bin/env python3

import json
from ref_helper import *

class Player():
    def __init__(self, server):
        self.wins = 0

        self.connection = Connection(server)
        self.name = self.connection.listen(1024).decode().rstrip()

    def addWin(self):
        self.wins += 1
        return self.wins

    def move(self):
        self.connection.send('move')
        move = self.connection.listen(1024).decode().rstrip()
        return move

class Game():

    def __init__(self, numPlayers):
        self.players = []

        #Create players
        for i in range(numPlayers):
            self.players.append(Player(playerServer))

        manager.send("match", "start")

        winner = self.runGame()
        manager.send("round", "end")
        self.reportResults("roundresult", winner)

        manager.send("match", "end")
        self.reportResults("matchresult", winner)

    def reportResults(self, resulttype, winner):
        for player in self.players:
            #Tell the player who won
            player.connection.send(winner.name + " wins\n")
            #Get the result for who won
            result = "Win" if (player.name == winner.name) else "Loss"
            manager.send(resulttype, [player.name, result, str(player.wins)])

    def runGame(self):
        #Loop until someone wins
        manager.send("round", ["start", 0])
        while True:
            for player in self.players:
                move = player.move()
                #Report move to manager
                manager.send("move", [player.name + " plays " + move, json.dumps([player.name, move])])
                #Check if it's a winning move
                if self.checkWin(move):
                    player.addWin()
                    return player

    def checkWin(self, move):
        return move == 'w'

game = Game(options.num)
